exonum/src/runtime/mailbox/mod.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright 2019 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! The module containing interfaces to request changes in the blockchain structure.
+
+mod blockchain_secretary;
+
+pub use blockchain_secretary::{BlockchainSecretary, MailboxContext};
+
+use std::cell::RefCell;
+
+use crate::runtime::{ArtifactId, ConfigChange, InstanceId, InstanceSpec};
+
+/// Optional callback to be called after request is completed by the blockchain core.
+pub type AfterRequestCompleted = Option<Box<dyn FnOnce() + 'static>>;
+
+/// An interface for runtimes to interact with the Exonum blockchain core.
+///
+/// All the requests added to the mailbox will be processed by the core `Blockchain` structure.
+///
+/// **Important note:** All the requests received after the transaction execution are considered
+/// **the part of execution process**. So, if service requests blockchain to perform some action
+/// and an error occurs during the request processing, the transaction will be treated as failed
+/// and, as a result, rolled back.
+///
+/// **Policy on request failures:**
```

Comment:
This is terribly complicated. Just sayin'.

Reply:
Do you have a proposal how to make it simpler, given the fact that requests will be processed when context in which a request was created is already gone?